### ISSUE TITLE
blobfuse2 mount fails with error Libfuse::initFuse : failed to mount fuse

### ISSUE THREAD

**Issue by ocoleman-cordel** [Azure/azure-storage-fuse#864]
### Which version of the blobfuse was used?
v2

### Which OS (please include version) are you using?
Ubuntu 20.04 (in docker container)

### What problem was encountered?
mount fails with error Libfuse::initFuse : failed to mount fuse

### Have you found a mitigation/solution?
No

### By default, blobfuse logs errors to syslog. If this is relevant, is there anything in the syslog that might be helpful?
```
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_CRIT [mount.go (296)]: Starting Blobfuse2 Mount : 2.0.0-preview.2 on (Ubuntu 20.04.3 LTS)
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_CRIT [mount.go (297)]: Logging level set to : LOG_INFO
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_INFO [libfuse.go (233)]: Libfuse::Configure : read-only false, allow-other true, default-perm 511, entry-timeout 120, attr-time 120, negative-timeout 120
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_INFO [file_cache.go (246)]: FileCache::Configure : Using default eviction policy
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_INFO [file_cache.go (255)]: FileCache::Configure : create-empty false, cache-timeout 120, tmp-path /workspace/temp/blobfusecache/axdevelblob_axml-olivercoleman-devfull_bash
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_INFO [config.go (298)]: BlockBlob::ParseAndValidateConfig : using the following proxy address from the config file: 
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_INFO [config.go (302)]: BlockBlob::ParseAndValidateConfig : sdk logging from the config file: false
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_INFO [config.go (367)]: ParseAndValidateConfig : Account: axdevelblob, Container: axml-olivercoleman-devfull, AccountType: BLOCK, Auth: KEY, Prefix: , Endpoint: https://axdevelblob.blob.core.windows.net/, ListBlock: 0
Tue Aug 16 06:55:27 UTC 2022 : blobfuse2[81] : LOG_INFO [config.go (371)]: ParseAndValidateConfig : Retry Config: Retry count 3, Max Timeout 3600, BackOff Time 1, Max Delay 3
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_CRIT [mount.go (296)]: Starting Blobfuse2 Mount : 2.0.0-preview.2 on (Ubuntu 20.04.3 LTS)
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_CRIT [mount.go (297)]: Logging level set to : LOG_INFO
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_INFO [libfuse.go (233)]: Libfuse::Configure : read-only false, allow-other true, default-perm 511, entry-timeout 120, attr-time 120, negative-timeout 120
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_INFO [file_cache.go (246)]: FileCache::Configure : Using default eviction policy
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_INFO [file_cache.go (255)]: FileCache::Configure : create-empty false, cache-timeout 120, tmp-path /workspace/temp/blobfusecache/axdevelblob_axml-olivercoleman-devfull_bash
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_INFO [config.go (298)]: BlockBlob::ParseAndValidateConfig : using the following proxy address from the config file: 
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_INFO [config.go (302)]: BlockBlob::ParseAndValidateConfig : sdk logging from the config file: false
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_INFO [config.go (367)]: ParseAndValidateConfig : Account: axdevelblob, Container: axml-olivercoleman-devfull, AccountType: BLOCK, Auth: KEY, Prefix: , Endpoint: https://axdevelblob.blob.core.windows.net/, ListBlock: 0
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_INFO [config.go (371)]: ParseAndValidateConfig : Retry Config: Retry count 3, Max Timeout 3600, BackOff Time 1, Max Delay 3
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_INFO [libfuse_handler.go (171)]: Libfuse::initFuse : Mounting with fuse3 library
Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_ERR [libfuse_handler.go (174)]: Libfuse::initFuse : failed to mount fuse
```

### If relevant, please share your mount command.
`blobfuse2 mount /workspace/data --config-file=/workspace/code/axlearning/config.yaml`

Config file:
```
allow-other: true

logging:
    type: base
    level: log_info
    file-path: /workspace/code/axlearning/bf2.log
file_cache:
    path: <redacted>
azstorage:
    endpoint: https://axdevelblob.blob.core.windows.net
    account-name: axdevelblob
    account-key: <redacted>
    mode: key
    container: <redacted>
components:
    - libfuse
    - file_cache
    - azstorage
```

**Comment by vibhansa-msft** [comment]
"Tue Aug 16 06:55:28 UTC 2022 : blobfuse2[93] : LOG_ERR [libfuse_handler.go (174)]: Libfuse::initFuse : failed to mount fuse
" : Indicates system is not able to mount using fuse3 driver. What linux distro and version you are using ? Suspecting some libfuse compatibility issue. Can you post output of "fusermount3 --version" command.

**Comment by ocoleman-cordel** [comment]
That was it, I had fuse v2.99 installed, the default version in Ubuntu 20.04. I've successfully mounted a blob container with blobfuse2.

While we're here: is it possible to specify all of the config options that can be defined in a config file via CLI options? For example `file_cache.max-size-mb`? We would rather not use config files because, or at least not config files that are only used by blobfuse, as we have our own config files for each environment that contain settings for blobfuse (and settings for other things too), and would like to keep all configuration in one config file.

**Comment by vibhansa-msft** [comment]
If your existing config files are in yaml format, then you shall be able to add blobfuse configs also to that file and provide that file as input to blobfuse.
Moving all configs from file to cli is not possible as we have a huge variety of config options. You can check this file for all possible config options: https://github.com/Azure/azure-storage-fuse/blob/main/setup/baseConfig.yaml

**Comment by XavierGeerinck** [comment]
Same issue here, `fusermount3 version: 3.9.0` and almost similar config file

**Comment by XavierGeerinck** [comment]
Solved, if you are running in a Docker container, start it up with: `--cap-add=SYS_ADMIN --device=/dev/fuse --security-opt apparmor:unconfined` (or just use `--privileged` but I am not a fan of that)

E.g.

```
sudo docker run -it --rm --cap-add=SYS_ADMIN --device=/dev/fuse --security-opt apparmor:unconfined azure-blobfuse /bin/bash
```

Working on a Blog Post, will try to link it here when finished

**Comment by vibhansa-msft** [comment]
Thanks @XavierGeerinck for your detailed blog. It will defiantly help a lot of users.

**Comment by vibhansa-msft** [comment]
Closing this as its working with suggested changes. Feel free to reopen if the issue persists or you have any further questions.

**Comment by dashesy** [comment]
I have this issue with Ubuntu 18.  By mistake I first installed the package for Ubuntu 20 first, could this be the reason?

**Comment by gapra-msft** [comment]
Hi @dashesy, yes this can happen if you use the Ubuntu 20 package on Ubuntu 18 since we build Ubn 20 for fuse3 and Ubn 18 for fuse2 (since Ubn 18 is not fuse3 compatible). Please install the Ubuntu 18 package on your machine if that is the version you are on.

**Comment by dashesy** [comment]
Thanks. I reinstalled the proper package and issue persisted. Then removed "read-only" snd "allow-other" from config, then it worked. Need to repro again but I think one of those has issue, on Ubuntu 18 on a vm

**Comment by gapra-msft** [comment]
@dashesy, have you added user_allow_other to your /etc/fuse.conf? That should make allow_other work

**Comment by dashesy** [comment]
No. Will try for repro. I didn't even want allow-other I just used the sample config.

**Comment by gdubya** [comment]
Adding `user_allow_other` fixed it for me. Thanks @gapra-msft !

**Comment by yusuf-jkhan1** [comment]
In case others have this issue:

I'm using `WSL2` with `Ubuntu 20.04`
I have `fusermount3 version: 3.9.0`

TL;DR: libfuse3 doesn't support the nonempty option anymore, remove it from the config.

I was getting this error:

```Tue Jan  3 12:06:00 PST 2023 : blobfuse2[20880] : LOG_DEBUG [libfuse_handler.go (224)]: Libfuse::populateFuseArgs : opts : -f
Tue Jan  3 12:06:00 PST 2023 : blobfuse2[20880] : LOG_DEBUG [libfuse_handler.go (224)]: Libfuse::populateFuseArgs : opts : -ofsname=blobfuse2
Tue Jan  3 12:06:00 PST 2023 : blobfuse2[20880] : LOG_DEBUG [libfuse_handler.go (224)]: Libfuse::populateFuseArgs : opts : -okernel_cache
Tue Jan  3 12:06:00 PST 2023 : blobfuse2[20880] : LOG_INFO [libfuse_handler.go (175)]: Libfuse::initFuse : Mounting with fuse3 library
Tue Jan  3 12:06:00 PST 2023 : blobfuse2[20880] : LOG_ERR [libfuse_handler.go (178)]: Libfuse::initFuse : failed to mount fuse
Tue Jan  3 12:06:00 PST 2023 : blobfuse2[20880] : LOG_ERR [libfuse.go (150)]: Libfuse::Start : Failed to init fuse [failed to mount fuse]
Tue Jan  3 12:06:00 PST 2023 : blobfuse2[20880] : LOG_ERR [mount.go (462)]: mount: error unable to start pipeline [failed to mount fuse]
```
This was my config file:
```
allow-other: true
nonempty: true

logging:
  type: syslog
  level: log_debug
  track-time: true

components:
  - libfuse
  - file_cache
  - attr_cache
  - azstorage

libfuse:
  attribute-expiration-sec: 120
  entry-expiration-sec: 120
  negative-entry-expiration-sec: 240

file_cache:
  path: ~/.blobfuse2/tmpcache
  timeout-sec: 120
  max-size-mb: 4096

attr_cache:
  timeout-sec: 7200

azstorage:
  type: block
  account-name: <redacted>
  account-key: <redacted>
  endpoint: https://<redacted>.blob.core.windows.net/
  mode: key
  container: <redacted>
```
This is my /etc/fuse.conf:

```
# The file /etc/fuse.conf allows for the following parameters:
#
# user_allow_other - Using the allow_other mount option works fine as root, in
# order to have it work as user you need user_allow_other in /etc/fuse.conf as
# well. (This option allows users to use the allow_other option.) You need
# allow_other if you want users other than the owner to access a mounted fuse.
# This option must appear on a line by itself. There is no value, just the
# presence of the option.

user_allow_other


# mount_max = n - this option sets the maximum number of mounts.
# Currently (2014) it must be typed exactly as shown
# (with a single space before and after the equals sign).
```
I allowed the user_allow_other option

Removing the `nonempty: true` from my config solved the issue.

**Comment by sytelus** [comment]
@ocoleman-cordel unfortunately doesn't give much detail how he fixed it so we are left hanging. In my case I did the followings (I'm not using docker).

1. `fusermount3 --version` print 3.9.0 and there seems to be nothing wrong with it.
2. Make sure all install step in Opton 1 was followed from https://learn.microsoft.com/en-us/azure/storage/blobs/blobfuse2-how-to-deploy#how-to-install-blobfuse2. Specifically, make sure you didn't skip configuring Microsoft repository step. This step installs libfuse and without it you will get above error.
3. Remove `nonempty: true` as discovered by @yusuf-jkhan1.
4. Make sure you run bobfuse2 with `sudo`. You need to use `sudo` for both mount and unmount commands.

**Comment by vibhansa-msft** [comment]
@sytelus: Thanks for putting up this list. Here are my points regarding this topic:

1. fusermount3 is a linux distro specific thing. Most latest distros support it but if you are using an older linux distro that lacks this support, you are on libfuse2 compatible system. From our end we cannot validate on every single linux distro and hence running this command is a way to identify what version of libfuse you are using.
2. If you install Blobfuse using the standard RPM/Deb packages published on the linux distros, it shall install compatible libfuse version on your system automatically.
3. Default logging level for blobfuse is WARNING, and that is enough for most common errors. When it says "failed to init fuse" error is coming from libfuse lib and not blobfuse code and hence there is nothing blobfuse can report here. This error message is printed in your log file even if log is at WARNING level. In such cases mostly its some user error with configuration like using allow_other but not enabling allow_other in libfuse settings.
4. nonempty: true is a discouraged feature and hence we do not mention it much in documents. We have seen customers having wrong expectations with this flag set so by default we do not allow nonempty mount.
5. mount and unmount does not need sudo. This purely depends on your system configuration. Again, nothing that is blocked from blobfuse end.
6. If there are errors on mount which are coming from azure storage then yes you will need to tweak the logging policy and get more logs to identify the root-cause of the error.

**Comment by sytelus** [comment]
@vibhansa-msft - it would be great to take customer centric view. Vast majority of folks are using debian/Ubuntu. Wouldn't it be good idea to validate dependency for this distro? 

Currently, bobfuse2 just silently fails. No error, nothing. How do you expect customers to figure out what went wrong? It doesn't even print "check logs". The bobfuse client can easily filter warnings and printed on stdio. It's never a great experience for any command line tool to just silently fail.

There is nowhere in documentation I saw `nonempty:true` is discouraged. In fact, there is no real documentation on config options to begin with. The only one I can find is [this](https://learn.microsoft.com/en-us/azure/storage/blobs/blobfuse2-configuration) which redirects to [this](https://github.com/Azure/azure-storage-fuse/blob/main/setup/baseConfig.yaml) on GitHub. And that has no information on not using `nonempty:true`.

I am using nothing fancy. Just run of the mill Ubuntu 20. Without `sudo` bobfuse2 just silently fails. There is no error why.

Other nitpick:
The [install instructions](https://learn.microsoft.com/en-us/azure/storage/blobs/blobfuse2-how-to-deploy#how-to-install-blobfuse2) should detect distro instead of hard coded URLs. That is, it would look like below:

```
distribution=$(. /etc/os-release;echo $ID/$VERSION_ID)
wget https://packages.microsoft.com/config/$distribution/packages-microsoft-prod.deb
sudo dpkg -i packages-microsoft-prod.deb
sudo apt-get update
sudo apt-get install libfuse3-dev fuse3
sudo apt-get install blobfuse2
```

There is too much boilerplate required in YAML because it doesn't use the principle of good defaults. Ideally, I shouldn't have to enter anything more than `azstorage` section and rest of the defaults should be good enough. But this isn't the case For example, `allow-non-empty-temp` must be true or bobfuse2 will silently fail on 2nd run.

I think bobfuse2 is very important tool in the world of AI and it would be great to make it super easy to use and more robust and pleasure to use.

**Comment by mickey-stringer** [comment]
@vibhansa-msft I am also having this issue and logs aren't giving me any clear info. Curiously, they show the mount is successful, but after a few more steps, the operation ends with the same "failed to mount fuse" error:
```Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (180)]: Libfuse::initFuse : Mounting with fuse3 library
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749194] : LOG_INFO [mount.go (465)]: mount: Child [3749201] mounted successfully at /sftpmount
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_TRACE [libfuse_handler.go (264)]: Libfuse::libfuse_init : init (read : 0, write 131072, read-ahead 131072)
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (266)]: Libfuse::NotifyMountToParent : Notifying parent for successful mount
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (273)]: Libfuse::libfuse_init : Kernel Caps : 2097115
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (280)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_PARALLEL_DIROPS
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (286)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_AUTO_INVAL_DATA
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (293)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_READDIRPLUS
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (299)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_ASYNC_READ
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (305)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_SPLICE_WRITE
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_INFO [libfuse_handler.go (317)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_WRITEBACK_CACHE
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_TRACE [libfuse_handler.go (340)]: Libfuse::libfuse_destroy : destroy
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_ERR [libfuse_handler.go (183)]: Libfuse::initFuse : failed to mount fuse
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_ERR [libfuse.go (166)]: Libfuse::Start : Failed to init fuse [failed to mount fuse]
Wed Jan 17 17:53:03 UTC 2024 : blobfuse2[3749201] : LOG_ERR [mount.go (548)]: mount: error unable to start pipeline [failed to mount fuse]
```

I'm running CentOS 8.2, I installed Blobfuse2 via the Microsoft repositories, and the output of `fusermount3 --version` is 3.2.1
Here is my config file (I do have `allow_user_other` set in /etc/fuse.conf):
```
# Common configurations
allow-other: true

logging:
  type: base
  level: log_debug
  file-path: /var/log/blobfuse2/blobfuse.log

# Pipeline configuration. Choose components to be engaged. The order below is the priority order that needs to be followed.
components:
  - libfuse
  - file_cache
  - azstorage

libfuse:
  default-permission: 0644
  fuse-trace: true

# Disk cache related configuration
file_cache:
  path: /var/tmp/blobfuse2

# Azure storage configuration
azstorage:
# Required
  type: block
  account-name: [redacted]
  container: sftpmount
  endpoint: [redacted]
  mode: key
  account-key: [redacted]
  virtual-directory: false
```

**Comment by vibhansa-msft** [comment]
What version of blobfuse2 you are using. Somehow logs that you have shared does not match with our latest version. If so, can you upgrade to latest version and then try again and share the logs if it continues to fail.

**Comment by mickey-stringer** [comment]
@vibhansa-msft I have 2.1.2-1 installed:
```
Installed Packages
blobfuse2.x86_64                             2.1.2-1                              @packages-microsoft-com-prod
```

**Comment by mickey-stringer** [comment]
@vibhansa-msft I started over with a fresh VM (still on CentOS 8.2), here's what I encountered at each step:
1. Configure Microsoft repository
2. Run `dnf install blobfuse2` and the following gets installed:
```
Installing:
 blobfuse2             x86_64           2.1.2-1                   packages-microsoft-com-prod            15 M
Installing dependencies:
 fuse-common           x86_64           3.2.1-12.el8              BaseOS-openlogic                       21 k
 fuse3                 x86_64           3.2.1-12.el8              BaseOS-openlogic                       50 k
```
3. Run `blobfuse2 mount /sftpmount --config-file=./bfconf.yaml` and get this error:
```
blobfuse2: error while loading shared libraries: libfuse3.so.3: cannot open shared object file: No such file or directory
```
4. Run `fusermount3 --version` and the output is 3.2.1
5. Run `dnf install libfuse3.so.3` (in reference to comments on [this issue](https://github.com/Azure/azure-storage-fuse/issues/1039)) and the following are installed/upgraded:
```
Installing:
 fuse3-libs                     i686              3.2.1-12.el8              BaseOS-openlogic            106 k
Upgrading:
 glibc                          x86_64            2.28-164.el8              BaseOS-openlogic            3.6 M
 glibc-all-langpacks            x86_64            2.28-164.el8              BaseOS-openlogic             25 M
 glibc-common                   x86_64            2.28-164.el8              BaseOS-openlogic            1.3 M
Installing dependencies:
 glibc                          i686              2.28-164.el8              BaseOS-openlogic            3.4 M
```
6. Re-run the mount command and get the same error about libfuse3.so.3:
```
blobfuse2: error while loading shared libraries: libfuse3.so.3: cannot open shared object file: No such file or directory
```
7. Run `dnf install fuse3-libs` (again in reference to [this issue](https://github.com/Azure/azure-storage-fuse/issues/1039)) and it installs the x86_64 build:
```
Installing:
 fuse3-libs              x86_64              3.2.1-12.el8                 BaseOS-openlogic               94 k
```
8. Re-run the mount command and it appears to work - the command runs and nothing is output to the shell. However, the logs show the same initial success and then failure:
```
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (180)]: Libfuse::initFuse : Mounting with fuse3 library
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_TRACE [libfuse_handler.go (264)]: Libfuse::libfuse_init : init (read : 0, write 131072, read-ahead 131072)
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (266)]: Libfuse::NotifyMountToParent : Notifying parent for successful mount
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (273)]: Libfuse::libfuse_init : Kernel Caps : 2097115
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (280)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_PARALLEL_DIROPS
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (286)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_AUTO_INVAL_DATA
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (293)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_READDIRPLUS
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (299)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_ASYNC_READ
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (305)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_SPLICE_WRITE
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_INFO [libfuse_handler.go (317)]: Libfuse::libfuse_init : Enable Capability : FUSE_CAP_WRITEBACK_CACHE
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23540] : LOG_INFO [mount.go (465)]: mount: Child [23547] mounted successfully at /sftpmount
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_TRACE [libfuse_handler.go (340)]: Libfuse::libfuse_destroy : destroy
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_ERR [libfuse_handler.go (183)]: Libfuse::initFuse : failed to mount fuse
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_ERR [libfuse.go (166)]: Libfuse::Start : Failed to init fuse [failed to mount fuse]
Thu Jan 18 17:36:21 UTC 2024 : blobfuse2[23547] : LOG_ERR [mount.go (548)]: mount: error unable to start pipeline [failed to mount fuse]
```

**Comment by vibhansa-msft** [comment]
Interesting, appears libfuse is failing the init for some reason. Can you enable fuse-trace and run in foreground. As a result you shall see some logs coming from libfuse on your console and that might give you hint on what libfuse did not like about the mount here.

**Comment by mickey-stringer** [comment]
@vibhansa-msft  that output this:
```
fuse: warning: library too old, some operations may not not work
FUSE library version: 3.2.1
nullpath_ok: 0
unique: 2, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
INIT: 7.31
flags=0x03fffffb
max_readahead=0x00020000
   INIT: 7.26
   flags=0x0001f039
   max_readahead=0x00400000
   max_write=0x00100000
   max_background=128
   congestion_threshold=96
   time_gran=1
   unique: 2, success, outsize: 80
fuse: reading device: Invalid argument
Error: unable to start pipeline [failed to mount fuse]
```

**Comment by vibhansa-msft** [comment]
"fuse: reading device: Invalid argument" appears some problem with your fuse driver. Can you uninstall the fuse/fuse3 lib and reinstall. I also see a warning on the top saying "fuse: warning: library too old, some operations may not not work".

**Comment by mickey-stringer** [comment]
@vibhansa-msft Okay I did that and still receive the same error. Just to be clear, this is a fresh VM with fresh installs of the latest available blobfuse2, fuse3, fuse3-libs, and libfuse3.so.3. 

I notice other users tend to be on Ubuntu and have different package versions available to them. Has blobfuse2 been validated on CentOS 8 (and with its available packages)? Or are there any known conflicts or prerequisites missing from the docs? My environment feels like a prototypical deployment, and yet I can't even get off the ground.

**Comment by vibhansa-msft** [comment]
We do run our test pipelines on CentOS 7.9 and 8.5. Binaries released for these versions are build and tested on those distro only. As you are getting some invalid argument from fuse end, I am not clear on what is going wrong. Ideally fuse should have given clear indication on what it did not like but here I do not even see those logs. Is it possible for you to try out on some other linux distro just to validate that this issue is very CentOS specific or your VM specific.

**Comment by mickey-stringer** [comment]
@vibhansa-msft I tried these steps again on a brand new CentOS 8.5 VM and received the same warnings and errors.
I created a new Ubuntu 20.04 VM and it worked fine. 
Can you try yourself to install and mount blobfuse2 on a fresh CentOS VM?

I know very little about fuse, but since I'm using newly created, out-of-the-box VMs, it strikes me that there is something missing from the CentOS installation/configuration instructions. Blobfuse2 installs fuse3 and fuse-common as dependencies, but then libfuse3.so.3 is still missing until manually installed. And then the foreground output from fuse says the library is too old, even though it's version 3.2.1. 

How were your build VMs set up?

**Comment by vibhansa-msft** [comment]
We have not observed this error in any customer setup in past. Generally guideline is to install blobfuse, which installs required dependency as well. If not, user can manually install libfuse3 on the system as well. I am not sure on why libfuse SO is missing in your case, fuse3 installation itself will bring that file to your system. Can you try with any other linux distro like Ubuntu or RHEL just to validate that this is very system specific issue you are hitting here.
